iked/vroute.c

```diff
@@ -203,21 +208,34 @@ vroute_insertroute(struct iked *env, int rdomain, struct sockaddr *dest)
 	if (route == NULL)
 		fatalx("%s: calloc.", __func__);
 
-	memcpy(&route->vr_dest, dest, dest->sa_len);
+	if (dest != NULL) {
+		route->vr_flags |= RTA_DST;
+		memcpy(&route->vr_dest, dest, dest->sa_len);
+	}
+	if (mask != NULL) {
+		route->vr_flags |= RTA_NETMASK;
+		memcpy(&route->vr_mask, mask, mask->sa_len);
+	}
 	route->vr_rdomain = rdomain;
 
 	TAILQ_INSERT_TAIL(&ivr->ivr_routes, route, vr_entry);
 }
 
 void
-vroute_removeroute(struct iked *env, int rdomain, struct sockaddr *dest)
+vroute_removeroute(struct iked *env, int rdomain, struct sockaddr *dest,
+    struct sockaddr *mask)
 {
 	struct iked_vroute_sc	*ivr = env->sc_vroute;
 	struct vroute_route	*route;
 
 	TAILQ_FOREACH(route, &ivr->ivr_routes, vr_entry) {
 		if (sockaddr_cmp(dest, (struct sockaddr *)&route->vr_dest, -1))
 			continue;
+		if (mask && !(route->vr_flags | RTA_NETMASK))
+			continue;
+		if (mask &&
+		    sockaddr_cmp(mask, (struct sockaddr *)&route->vr_mask, -1))
```

Comment:
Usually I'd put the call in the first line, and one of the arguments in the second, but maybe what you did is better for readability.